openwisp_users/utils.py

```diff
@@ -0,0 +1,44 @@
+from django.conf import settings
+
+if 'reversion' in settings.INSTALLED_APPS:  # pragma: no cover
+    from reversion.admin import VersionAdmin as BaseModelAdmin
+else:
+    from django.contrib.admin import ModelAdmin as BaseModelAdmin
+
+
+class BaseAdmin(BaseModelAdmin):
+    history_latest_first = True
+
+
+def usermodel_add_form(model, additional_fields):
+    for field in additional_fields:
+        modelMeta = model.add_form.Meta
+        # Add form fieldsets
+        add_fieldsets = modelMeta.fieldsets[0][1]['fields'][:]
+        modelMeta.fieldsets[0][1]['fields'] = (
+            add_fieldsets[: field[0]] + [field[1]] + add_fieldsets[field[0] :]
+        )
+        # Add form fieldsets_superuser
+        add_fieldsets_superuser = modelMeta.fieldsets_superuser[0][1]['fields'][:]
+        modelMeta.fieldsets_superuser[0][1]['fields'] = (
+            add_fieldsets_superuser[: field[0]]
+            + [field[1]]
+            + add_fieldsets_superuser[field[0] :]
+        )
+
+
+def usermodel_change_form(model, additional_fields):
+    # Change form fieldsets
+    for field in additional_fields:
+        fieldsets = model.fieldsets[1][1]['fields'][:]
+        model.fieldsets[1][1]['fields'] = (
+            fieldsets[: field[0]] + [field[1]] + fieldsets[field[0] :]
+        )
+
+
+def usermodel_list_and_search(model, additional_fields):
+    # Change form fieldsets
+    for field in additional_fields:
+        displays = model.list_display[:]
+        model.list_display = displays[: field[0]] + [field[1]] + displays[field[0] :]
+        model.search_fields += (field[1],)
```

Comment:
maybe it would help if this file was in base/admin.py?

Reply:
I've documented them in the README : https://github.com/openwisp/openwisp-users/pull/123/files#diff-88b99bb28683bd5b7e3a204826ead112R509-R527

I'll link that documentation here for the developers as well! :smile:

Reply:
> maybe it would help if this file was in base/admin.py?

It's not really used in the `base/admin.py` and is optional to be used when extending, hence I added it here.

Should I have this and 

```
from django.conf import settings

if 'reversion' in settings.INSTALLED_APPS:  # pragma: no cover
    from reversion.admin import VersionAdmin as BaseModelAdmin
else:
    from django.contrib.admin import ModelAdmin as BaseModelAdmin


class BaseAdmin(BaseModelAdmin):
    history_latest_first = True
```

with it or just the new functions?

Reply:
Ok cool, let's not move it, please just a brief comment regarding what this file is for, like you did in the README but shorted or point to the README.

Reply:
Already done, there is a doc string in the functions, which points to the README like:
```
    """
    Read:
    https://github.com/openwisp/openwisp-users/blob/master/README.rst#usermodel_add_form
    """
```